core/lib/Cardano/Address/KeyHash.hs

```diff
@@ -0,0 +1,313 @@
+              Just (CommitteeCold, bytes)
+        | hrp == CIP5.cc_cold_vkh && checkBSLength bytes 28 =
+              Just (CommitteeCold, bytes)
+        | hrp == CIP5.cc_hot && checkBSLength bytes 29 =
+              let (fstByte, payload) = first BS.head $ BS.splitAt 1 bytes
+              --   hot           0000....
+              --   keyhash       ....0010
+              in if fstByte == 0b00000010 then
+                  Just (CommitteeHot, payload)
+                 else
+                  Nothing
+        | hrp == CIP5.cc_hot && checkBSLength bytes 28 =
+              Just (CommitteeHot, bytes)
+        | hrp == CIP5.cc_hot_vkh && checkBSLength bytes 28 =
+              Just (CommitteeHot, bytes)
+        | hrp == CIP5.addr_shared_vk && checkBSLength bytes 32 =
+              Just (Payment, hashCredential bytes)
+        | hrp == CIP5.addr_vk && checkBSLength bytes 32 =
+              Just (Payment, hashCredential bytes)
+        | hrp == CIP5.addr_shared_xvk && checkBSLength bytes 64 =
+              Just (Payment, hashCredential $ BS.take 32 bytes)
+        | hrp == CIP5.addr_xvk && checkBSLength bytes 64 =
+              Just (Payment, hashCredential $ BS.take 32 bytes)
+        | hrp == CIP5.stake_shared_vk && checkBSLength bytes 32 =
+              Just (Delegation, hashCredential bytes)
+        | hrp == CIP5.stake_vk && checkBSLength bytes 32 =
+              Just (Delegation, hashCredential bytes)
+        | hrp == CIP5.stake_shared_xvk && checkBSLength bytes 64 =
+              Just (Delegation, hashCredential $ BS.take 32 bytes)
+        | hrp == CIP5.stake_xvk && checkBSLength bytes 64 =
+              Just (Delegation, hashCredential $ BS.take 32 bytes)
+        | hrp == CIP5.policy_vk && checkBSLength bytes 32 =
+              Just (Policy, hashCredential bytes)
+        | hrp == CIP5.policy_xvk && checkBSLength bytes 64 =
+              Just (Policy, hashCredential $ BS.take 32 bytes)
+        | hrp == CIP5.drep_vk && checkBSLength bytes 32 =
+              Just (Representative, hashCredential bytes)
+        | hrp == CIP5.drep_xvk && checkBSLength bytes 64 =
+              Just (Representative, hashCredential $ BS.take 32 bytes)
+        | hrp == CIP5.cc_cold_vk && checkBSLength bytes 32 =
+              Just (CommitteeCold, hashCredential bytes)
+        | hrp == CIP5.cc_cold_xvk && checkBSLength bytes 64 =
+              Just (CommitteeCold, hashCredential $ BS.take 32 bytes)
+        | hrp == CIP5.cc_hot_vk && checkBSLength bytes 32 =
+              Just (CommitteeHot, hashCredential bytes)
+        | hrp == CIP5.cc_hot_xvk && checkBSLength bytes 64 =
+              Just (CommitteeHot, hashCredential $ BS.take 32 bytes)
+        | otherwise = Nothing
+    checkBSLength :: ByteString -> Int -> Bool
+    checkBSLength bytes expLength =
+        BS.length bytes == expLength
+
+-- Possible errors when deserializing a key hash from text.
+--
+-- @since 3.0.0
+data ErrKeyHashFromText
+    = ErrKeyHashFromTextInvalidString
```

Comment:
Ideally, I would add a payload to each errors to better pinpoint the source of the error and help the user fix it.

Reply:
I added some, but will revisit this also in the future